File: labs/control-ui/api-shadow/edge_cdn_pool_sorter.ts
export interface EdgeIpStatus {
  ipAddress: string;
  latencyMs: number;
  packetLossRatio: number;
  isAvailable: boolean;
}

export class EdgeCdnPoolSorter {
  private ipPool = new Map<string, EdgeIpStatus>();
  private maxLatencyThresholdMs: number;
  constructor(maxLatencyThresholdMs: number = 300) {
    this.maxLatencyThresholdMs = maxLatencyThresholdMs;
  }

  addIp(ip: string): void {
    this.ipPool.set(ip, {
      ipAddress: ip,
      latencyMs: Infinity,
      packetLossRatio: 1.0,
      isAvailable: false,
    });
  }

  updateProbeResult(ip: string, latencyMs: number, success: boolean): void {
    const entry = this.ipPool.get(ip);
    if (!entry) return;

    if (success) {
      entry.latencyMs = latencyMs;
      entry.packetLossRatio = 0.0;
      entry.isAvailable = latencyMs <= this.maxLatencyThresholdMs;
    } else {
      entry.packetLossRatio = 1.0;
      entry.isAvailable = false;
    }
  }

  getSortedFastest(): EdgeIpStatus[] {
    return Array.from(this.ipPool.values())
      .filter((e) => e.isAvailable)
      .sort((a, b) => a.latencyMs - b.latencyMs);
  }

  bestIp(): string | undefined {
    return this.getSortedFastest()[0]?.ipAddress;
  }
}

File: labs/mobile/donor-archive/tunnel/EdgeCdnPoolSorter.kt
package com.luminet.android.tunnel

data class AndroidEdgeIpRecord(
    val ipAddress: String,
    var latencyMs: Long = Long.MAX_VALUE,
    var packetLossRatio: Float = 1.0f,
    var isAvailable: Boolean = false
)

class EdgeCdnPoolSorter(private val maxLatencyThresholdMs: Long = 300) {
    private val ipPool = mutableMapOf<String, AndroidEdgeIpRecord>()

    fun addIp(ip: String) {
        ipPool[ip] = AndroidEdgeIpRecord(ipAddress = ip)
    }

    fun updateProbeResult(ip: String, latencyMs: Long, success: Boolean) {
        val entry = ipPool[ip] ?: return
        if (success) {
            entry.latencyMs = latencyMs
            entry.packetLossRatio = 0.0f
            entry.isAvailable = latencyMs <= maxLatencyThresholdMs
        } else {
            entry.packetLossRatio = 1.0f
            entry.isAvailable = false
        }
    }

    fun getSortedFastest(): List<AndroidEdgeIpRecord> {
        return ipPool.values
            .filter { it.isAvailable }
            .sortedBy { it.latencyMs }
    }

    fun bestIp(): String? {
        return getSortedFastest().firstOrNull()?.ipAddress
    }
}
